### ISSUE TITLE
Unable to build

### ISSUE THREAD

**Issue by ickk** [sifive/freedom-tools#3]
Hello,

Following the directions given in the Readme.md I wasn't able to build the tools.

The exact commands I ran were:
```
sudo apt-get install cmake autoconf automake autotools-dev curl libmpc-dev libmpfr-dev libgmp-dev gawk build-essential bison flex texinfo gperf patchutils bc zlib1g-dev libexpat-dev libtool pkg-config mingw-w64
git clone --recurse-submodules https://github.com/sifive/freedom-tools.git
cd freedom-tools
make > make.stdout 2> make.stderr
```
the contents of `make.stderr` are:
```
/bin/sh: 1: ./configure: not found
make[1]: *** No rule to make target 'buildlib'.  Stop.
make[1]: *** No rule to make target 'installlib'.  Stop.
autoreconf: Entering directory `.'
autoreconf: configure.ac: not using Gettext
autoreconf: running: aclocal --force 
aclocal: warning: couldn't open directory 'm4': No such file or directory
autoreconf: configure.ac: tracing
autoreconf: running: libtoolize --copy --force
autoreconf: running: /usr/bin/autoconf --force
autoreconf: running: /usr/bin/autoheader --force
autoreconf: configure.ac: not using Automake
autoreconf: Leaving directory `.'
2019-06-09 03:55:14 URL: ftp://gcc.gnu.org/pub/gcc/infrastructure/gmp-6.1.0.tar.bz2 [2383840] -> "./gmp-6.1.0.tar.bz2" [1]
2019-06-09 03:55:19 URL: ftp://gcc.gnu.org/pub/gcc/infrastructure/mpfr-3.1.4.tar.bz2 [1279284] -> "./mpfr-3.1.4.tar.bz2" [1]
2019-06-09 03:55:23 URL: ftp://gcc.gnu.org/pub/gcc/infrastructure/mpc-1.0.3.tar.gz [669925] -> "./mpc-1.0.3.tar.gz" [1]
2019-06-09 03:55:28 URL: ftp://gcc.gnu.org/pub/gcc/infrastructure/isl-0.18.tar.bz2 [1658291] -> "./isl-0.18.tar.bz2" [1]
make[1]: *** No targets specified and no makefile found.  Stop.
make[1]: *** No rule to make target 'install'.  Stop.
make[1]: *** No rule to make target 'all-gcc'.  Stop.
make[1]: *** No rule to make target 'install-gcc'.  Stop.
make[1]: *** No targets specified and no makefile found.  Stop.
make[1]: *** No rule to make target 'install'.  Stop.
make[1]: *** No rule to make target 'install-pdf'.  Stop.
make[1]: *** No targets specified and no makefile found.  Stop.
make[1]: *** No rule to make target 'install'.  Stop.
find: ‘/home/nick/freedom-tools/obj/x86_64-linux-ubuntu14/build/riscv-gnu-toolchain/build-newlib-nano-install/riscv64-unknown-elf/lib’: No such file or directory
find: ‘/home/nick/freedom-tools/obj/x86_64-linux-ubuntu14/build/riscv-gnu-toolchain/build-newlib-nano-install/riscv64-unknown-elf/lib’: No such file or directory
find: ‘/home/nick/freedom-tools/obj/x86_64-linux-ubuntu14/build/riscv-gnu-toolchain/build-newlib-nano-install/riscv64-unknown-elf/lib’: No such file or directory
cp: cannot stat '/home/nick/freedom-tools/obj/x86_64-linux-ubuntu14/build/riscv-gnu-toolchain/build-newlib-nano-install/riscv64-unknown-elf/include/newlib.h': No such file or directory
/bin/sh: 4: cannot create obj/x86_64-linux-ubuntu14/build/riscv-gnu-toolchain/build-newlib-nano-install/stamp: Directory nonexistent
make: *** [Makefile:356: obj/x86_64-linux-ubuntu14/build/riscv-gnu-toolchain/build-newlib-nano-install/stamp] Error 2
configure: WARNING: If you wanted to set the --build type, don't use --host.
    If a cross compiler is detected then cross compile mode will be used.
configure: WARNING: If you wanted to set the --build type, don't use --host.
    If a cross compiler is detected then cross compile mode will be used.
configure: WARNING: If you wanted to set the --build type, don't use --host.
    If a cross compiler is detected then cross compile mode will be used.
configure: WARNING: using in-tree isl, disabling version check
*** This configuration is not supported in the following subdirectories:
     libdecnumber sim gdb
    (Any other directories should still work fine.)
*** This configuration is not supported in the following subdirectories:
     target-libquadmath target-libatomic target-libitm target-libsanitizer target-libvtv target-libmpx gnattools gotools target-libada target-libhsail-rt target-libgfortran target-libbacktrace target-libgo target-libffi target-libobjc target-libgomp target-liboffloadmic target-libssp
    (Any other directories should still work fine.)
```

My system is `Pop!_OS 18.10 x86_64`, which shouldn't differ meaningfully compared to Ubuntu 18.10.

I realise there are prebuilt binaries available, but I'd like to solve this issue if possible.

**Comment by jim-wilson** [comment]
Some of these errors are expected and harmless.  Some of them are not.  Having the stderr and stdout separated means all of the errors are missing the context and I can't see what went wrong.  It is better to pipe them into the same file.  You might also try using -j1 to get a clearer build log.

I'm guessing the first error happens in the riscv-openocd dir, since most other tools come with a configure script pre-built.  The build process creates a bunch of log files to help debug build failures.  Try looking at obj/x86_64-linux-ubuntu14/build/riscv-openocd/make-bootstrap.log.  I think you might find an error in there.  Maybe there is an autoconf/etc version mismatch problem here that prevents the configure program from being created.

**Comment by ickk** [comment]
Hey thanks for taking the time to help me with this. The `obj/x86_64-linux-ubuntu14/build/riscv-openocd` directory doesn't exist for me.

So just to have a clean slate I created a fresh VM with Ubuntu 16.04, installed the mentioned dependencies, cloned the repo, and tried `make -j1`; The build still failed to produce a `bin` directory or any tarballs. 

I've included the full output from make in this gist, from both the Ubuntu 16.04 virtual machine and my Pop!_OS 18.10 system: https://gist.github.com/ickk/0911e73970facfbf198357ab9f4adf77

**Comment by jim-wilson** [comment]
The first failure is at the top.

rm -rf obj/x86_64-linux-ubuntu14/build/expat/
mkdir -p obj/x86_64-linux-ubuntu14/build/expat/
cp -a src/libexpat/expat/* obj/x86_64-linux-ubuntu14/build/expat/
cd obj/x86_64-linux-ubuntu14/build/expat/; ./buildconf.sh &>make-buildconf.log
touch -c obj/x86_64-linux-ubuntu14/build/expat/configure
mkdir -p obj/x86_64-linux-ubuntu14/build/expat
cd obj/x86_64-linux-ubuntu14/build/expat; ./configure --prefix=/home/nick/freedom-tools/obj/x86_64-linux-ubuntu14/install/riscv64-unknown-elf-gcc-8.2.0-2019.02.0-x86_64-linux-ubuntu14 --host=x86_64-linux-gnu &>make-configure.log
/bin/sh: 1: ./configure: not found

So look in obj/x86_64-linux-ubuntu14/build/expat/make-buildconf.log file to see what went wrong.  As before, I'd guess that something is missing on your build machine.

Odd that the build did not fail immediately.  Are you perhaps using make -i?  That is a bad idea in general, as you will waste a lot of time building stuff you can't use, and will get a lot of cascading errors that will just confuse the situation.

**Comment by ickk** [comment]
`obj/x86_64-linux-ubuntu14/build/expat/make-buildconf.log` is an empty file. In fact all of the other log files in the `../build/expat` directory, `make-buildlib.log`, `make-configure.log`, `make-installlib.log`, are empty. - I'm not sure how `make-buildconf.log` is empty when manually running `cd obj/x86_64-linux-ubunutu14/build/expat/; ./buildconf.sh` does produce some output.

I am not using `make -i`. Originally I was using simply `make`, and after you suggested the single threaded option `make -j1` (the outputs in the gist I linked were both using make -j1).

I'm confused how something could be missing from the build machine when two completely different environments (16.04 guest virtual machine and 18.10 host) are seemingly failing in a similar way.

**Comment by jim-wilson** [comment]
A file could be empty if the disk was full.  But then nothing would build if you have a full disk so that seems unlikely.

A file could be empty if /bin/sh is broken, or maybe does not support the syntax used in the makefile.  A broken /bin/sh could also explain why make doesn't exit on the first error, if /bin/sh is not returning program exit codes correctly.  This is also unlikely, but could be an issue if you have a /bin/sh that isn't pointing at bash, and there is bash specific syntax in the Makefile.  This is at least worth checking.  Try switching /bin/sh to bash if it isn't already.  If that works, then this could be considered a Makefile bug, as it should force use of bash if bash is required.

Another possible issue is that you have unusual environment variable settings.  Setting MAKEFLAG to include -i will give the same result as running "make -i" for instance.  There could also be other environment variables set that are confusing other parts of the build process.  This is also unlikely, but I don't have anything better to suggest, so it is worth checking your environment and getting rid of stuff you don't need that might confuse a build.

**Comment by ickk** [comment]
Both my build systems ship with `/bin/sh` linked to dash; changing to bash as you suggested helped the build proceed further.

I did hit a missing dependency for `.../riscv-gnu-toolchain/build-binutils-newlib` which required TeX - fixed with `apt install texlive`

I had an issue at `.../riscv-gnu-toolchain/build-gcc-newlib-stage1/`. The tail of `.../build-gcc-newlib-stage1/make-build.log` looked like
```
checking dynamic linker characteristics... configure: error: Link tests are not allowed after GCC_NO_EXECUTABLES.
make[1]: *** [Makefile:10193: configure-zlib] Error 1
make[1]: Leaving directory '/home/nick/freedom-tools/obj/x86_64-linux-ubuntu14/build/riscv-gnu-toolchain/build-gcc-newlib-stage1'
```
This same thing happened in both my build environments, so just as a stab in the dark on my part, I changed the top level Makefile at line 139: `$(UBUNTU64)-rgcc-configure   := --without-system-zlib` to `--with-system-zlib` to match the corresponding `DARWIN` and `REDHAT` lines, which I guess just bypassed building zlib(?).

The last remaining issue I had was another missing dependency of libusb for riscv-openocd - solved with `apt install libusb-1.0-0-dev`.

With that I am able to complete a build. Thanks for your patience helping me troubleshoot.
